specs/agents/tracing-spans-discarding.md

```diff
@@ -0,0 +1,128 @@
+# Discarding fast spans
+
+Currently,
+the experience of our APM solution is not ideal when monitoring an application that ends up generating thousands of spans per request.
+
+We currently have a default limit of 500 spans in the agents and 1000 spans in the UI to prevent overwhelming the system with too many spans.
+However,
+that just cuts off at a certain point without considering which spans are important and which are not.
+
+## `span_min_duration` configuration
+
+Sets the minimum duration of spans.
+Spans that execute faster than this threshold are attempted to be discarded.
+
+The attempt fails if they lead up to a span that can't be discarded.
+Spans that propagate the trace context to downstream services,
+such as outgoing HTTP requests,
+can't be discarded.
+Additionally, spans that lead to an error or that may be a parent of an async operation can't be discarded.
+
+However, external calls that don't propagate context,
+such as calls to a database, can be discarded using this threshold.
```

Comment:
In my opinion we should keep database calls even if they are below the threshold for the fast spans.

Reply:
I heard some feedback from users where they wanted a similar feature and in those cases it was all database spans. ORM frameworks sometimes create multiple db calls and in general I had the impression that db will be the one causing "span explosion" a lot. I feel not including db spans here would limit the usefulness of this config. 

But good [point about the service map](https://github.com/elastic/apm/pull/314#issuecomment-700001023) - is is unacceptable to just say that this influences servicemap?

Reply:
It was an explicit design goal and a request from you know who to reduce the vast amount of short DB calls from the waterfall.

I see that this can be problematic when trying to determine metrics for the service map. But then we might have to tackle the issue in a completely different way.
For example similar to how the RUM agent does it and fold multiple repetitive spans that have the same context into one and display as `10 x SELECT FROM users`, for example.
Or we have to track the metrics in the agent.

Reply:
@felixbarny :) I now remember that request. 

We should factor in a way to get the Database connections right. The Service Maps should show accurate connectivity data regardless of how fast it is being executed. In the future, we'd also want to add a dependencies/Database page to track all the database queries for a particular service - if we drop these spans now we will have to add back some logic in to add these back. Also, omitting fast SQL queries may hinder a user from detecting a n+1 query problem all together(especially if those queries are quick). 

I like the example you provided to fold multiple repetitive spans with the same context to display them. That would solve both the too many spans issue while keeping the maps accurate and allowing for any future development.

Reply:
Very good points, Neha. span_min_duration seems to be more of a band-aid fix. Re-implementing that for other agents probably doesn't make sense as it's not a good solution long-term.
Maybe we need a solution that tackles the problem of excessive data collection more holistically. Some thoughts on this:
Spans that exceed `transaction_max_spans` are currently discarded the same way as spans faster than `span_min_duration`. Maybe we need a way of attaching information to the transaction about how many spans were discarded by span type and destination. We could use this information both in the waterfall view (we've omitted 42 DB spans to a MySQL DB that took 1.234s in total) and to calculate edge metrics to DBs accurately in case we dropped data. We already collect breakdown metrics that contain similar data but we don't group them by destination and don't attach them to transactions.
Another thing to reconsider is trying to deduplicate spans that just contain repetitive information, such as n+1 queries or 100s of Redis cache `GET` calls in a row.

Do you have a sense of the priority of this? Should we remove `span_min_duration` from 7.11? Should we try to get to a holistic solution within the 7.11 timeframe?